File: kotlin/src/main/kotlin/aethernet/handshake/HandshakeService.kt
// SPDX-License-Identifier: MIT

package aethernet.handshake

import aethernet.AetherNetConstants
import aethernet.protocol.MeshPacket
import aethernet.protocol.PacketType
import aethernet.routing.MeshSender
import org.slf4j.LoggerFactory
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

/**
 * Default capability handshake service. Tracks the peers we've Hello'd, the
 * peers we've finished negotiating with, and emits callbacks on completion /
 * incompatibility.
 *
 * Wire flow:
 *
 * ```
 * A → B   Hello       { min:1, max:2, caps:[X,Y,Z], impl:"…" }
 * A ← B   HelloAck    { min:1, max:2, caps:[X,Y],   impl:"…" }
 * ```
 *
 * Negotiation rules:
 *  - Negotiated version = `min(ourMax, theirMax)`.
 *  - If `min(ourMax,theirMax) < max(ourMin,theirMin)` the ranges do not
 *    overlap → fire [onIncompatiblePeer], refuse to lock in.
 *  - Locked-in capability set = `ourCaps ∩ theirCaps`.
 *
 * Mirrors `AetherNet.Handshake.HandshakeService` (C#) one-for-one. The JSON
 * shape on the wire is identical; cross-language interop is verified by C#
 * peers deserialising Kotlin-emitted Hello packets and vice versa.
 */
class HandshakeService(
    private val sender: MeshSender,
    private val ourMinVersion: Byte = 1,
    private val ourMaxVersion: Byte = AetherNetConstants.PROTOCOL_VERSION_CURRENT.toByte(),
    private val ourCapabilities: Set<String> = DEFAULT_CAPABILITIES,
    private val ourImplementation: String = DEFAULT_IMPLEMENTATION,
) {
    init {
        require(ourMinVersion <= ourMaxVersion) {
            "ourMinVersion ($ourMinVersion) cannot exceed ourMaxVersion ($ourMaxVersion)."
        }
    }

    private val logger = LoggerFactory.getLogger(HandshakeService::class.java)

    // Peers we've already sent a Hello to, to suppress duplicate sends.
    private val helloSent = ConcurrentHashMap.newKeySet<String>()

    // Peers we've finished negotiating with.
    private val negotiated = ConcurrentHashMap<String, PeerCapabilities>()

    /**
     * Invoked when negotiation completes (either via HelloAck receipt or via
     * the backward-compat fallback). Receives the locked-in [PeerCapabilities].
     */
    var onPeerNegotiated: ((PeerCapabilities) -> Unit)? = null

    /**
     * Invoked when a peer's announced version range does not overlap with
     * ours — we cannot speak to them. Subscribers should drop the peer from
     * their connected-peer set.
     */
    var onIncompatiblePeer: ((IncompatiblePeerEvent) -> Unit)? = null

    /**
     * Initiate a Hello toward a freshly-discovered peer. No-op if a Hello
     * has already been sent to this peer in the current session
     * (re-broadcasts can otherwise cause duplicate Hellos).
     */
    suspend fun initiate(peerUhid: String) {
        require(peerUhid.isNotEmpty()) { "peerUhid must not be empty" }
        if (peerUhid == sender.localUhid) return

        // Suppress duplicate Hellos.
        if (!helloSent.add(peerUhid)) return

        val packet = buildPacket(PacketType.Hello, peerUhid)
        val delivered = sender.send(packet, peerUhid)
        logger.debug("Hello sent to {} delivered={}", peerUhid, delivered)
    }

    /**
     * Handle an inbound [PacketType.Hello]: lock in the announced capabilities
     * and reply with a HelloAck.
     */
    suspend fun handleHello(helloPacket: MeshPacket) {
        require(helloPacket.type == PacketType.Hello) {
            "Expected Hello, got ${helloPacket.type}"
        }
        if (helloPacket.sourceUhid.isEmpty()) return
        if (helloPacket.sourceUhid == sender.localUhid) return

        val theirs = HelloPayload.fromJsonBytesOrNull(helloPacket.payload)
        if (theirs == null) {
            logger.warn("Hello from {} has malformed payload — ignoring", helloPacket.sourceUhid)
            return
        }

        val negotiatedRecord = tryNegotiate(helloPacket.sourceUhid, theirs) ?: return

        negotiated[helloPacket.sourceUhid] = negotiatedRecord
        onPeerNegotiated?.invoke(negotiatedRecord)
        logger.info(
            "Hello accepted from {} → version={} caps=[{}] impl={}",
            helloPacket.sourceUhid,
            negotiatedRecord.negotiatedVersion,
            negotiatedRecord.capabilities.joinToString(","),
            negotiatedRecord.implementationVersion,
        )

        // Reply with HelloAck — even if we already sent them an unprompted
        // Hello, the spec is symmetric and the ack carries our own range / caps.
        val ack = buildPacket(PacketType.HelloAck, helloPacket.sourceUhid)
        val delivered = sender.send(ack, helloPacket.sourceUhid)
        logger.debug("HelloAck sent to {} delivered={}", helloPacket.sourceUhid, delivered)
    }

    /**
     * Handle an inbound [PacketType.HelloAck]: lock in the negotiated
     * capabilities for the replying peer.
     */
    fun handleHelloAck(helloAckPacket: MeshPacket) {
        require(helloAckPacket.type == PacketType.HelloAck) {
            "Expected HelloAck, got ${helloAckPacket.type}"
        }
        if (helloAckPacket.sourceUhid.isEmpty()) return
        if (helloAckPacket.sourceUhid == sender.localUhid) return

        val theirs = HelloPayload.fromJsonBytesOrNull(helloAckPacket.payload)
        if (theirs == null) {
            logger.warn("HelloAck from {} has malformed payload — ignoring", helloAckPacket.sourceUhid)
            return
        }

        val negotiatedRecord = tryNegotiate(helloAckPacket.sourceUhid, theirs) ?: return

        negotiated[helloAckPacket.sourceUhid] = negotiatedRecord
        onPeerNegotiated?.invoke(negotiatedRecord)
        logger.info(
            "HelloAck received from {} → version={} caps=[{}] impl={}",
            helloAckPacket.sourceUhid,
            negotiatedRecord.negotiatedVersion,
            negotiatedRecord.capabilities.joinToString(","),
            negotiatedRecord.implementationVersion,
        )
    }

    /**
     * Look up the locked-in capabilities for a peer. Returns null if the
     * handshake has not yet completed — callers can either wait for the
     * [onPeerNegotiated] callback or proceed with caution.
     */
    fun getPeerCapabilities(peerUhid: String): PeerCapabilities? {
        require(peerUhid.isNotEmpty()) { "peerUhid must not be empty" }
        return negotiated[peerUhid]
    }

    /**
     * Drop a peer's cached capabilities and re-issue a Hello on the next
     * outbound contact. Used when version-mismatch is detected in subsequent
     * traffic.
     */
    fun renegotiate(peerUhid: String) {
        require(peerUhid.isNotEmpty()) { "peerUhid must not be empty" }
        negotiated.remove(peerUhid)
        helloSent.remove(peerUhid)
        logger.info("Cleared cached capabilities for {}; next contact will re-Hello", peerUhid)
    }

    /**
     * Snapshot of every peer that has finished negotiating, for diagnostics
     * / health-check use.
     */
    fun getAllNegotiated(): List<PeerCapabilities> = negotiated.values.toList()

    /**
     * Backward-compat: install a "v1, no caps" record for a peer that never
     * replied to our Hello within the timeout window. Hosts call this from
     * their own timer / heartbeat loop. Idempotent — if the peer has since
     * replied with a HelloAck, the existing record wins.
     */
    fun assumeLegacyV1(peerUhid: String) {
        require(peerUhid.isNotEmpty()) { "peerUhid must not be empty" }
        if (peerUhid == sender.localUhid) return

        val fallback = PeerCapabilities(
            peerUhid = peerUhid,
            negotiatedVersion = 1,
            capabilities = emptySet(),
            implementationVersion = "",
            negotiatedAt = Instant.now(),
        )

        // putIfAbsent — keep the existing record if a real HelloAck has
        // already populated one.
        val existing = negotiated.putIfAbsent(peerUhid, fallback)
        if (existing == null) {
            onPeerNegotiated?.invoke(fallback)
            logger.warn(
                "No HelloAck from {} after timeout — assuming protocol v1 / no advertised capabilities",
                peerUhid,
            )
        }
    }

    private fun buildPacket(type: PacketType, destinationUhid: String): MeshPacket {
        val payload = HelloPayload(
            minVersion = ourMinVersion,
            maxVersion = ourMaxVersion,
            capabilities = ourCapabilities.toList(),
            implementation = ourImplementation,
        )

        return MeshPacket(
            type = type,
            sourceUhid = sender.localUhid,
            destinationUhid = destinationUhid,
            ttl = 1, // direct hop only — handshake never relays
            priority = 0,
            protocolVersion = ourMaxVersion,
            payload = payload.toJsonBytes(),
        )
    }

    /**
     * Returns the negotiated record on success, or null and fires
     * [onIncompatiblePeer] on version-overlap failure.
     */
    private fun tryNegotiate(peerUhid: String, theirs: HelloPayload): PeerCapabilities? {
        if (theirs.minVersion > theirs.maxVersion) {
            logger.warn(
                "Handshake from {} announces inverted range min={} > max={} — refusing",
                peerUhid, theirs.minVersion, theirs.maxVersion,
            )
            fireIncompatible(peerUhid, theirs, "inverted version range")
            return null
        }

        // Overlap check: highest min must be ≤ lowest max.
        val overlapMin = maxOf(ourMinVersion.toInt(), theirs.minVersion.toInt())
        val overlapMax = minOf(ourMaxVersion.toInt(), theirs.maxVersion.toInt())
        if (overlapMin > overlapMax) {
            fireIncompatible(
                peerUhid, theirs,
                "no version overlap (ours=$ourMinVersion..$ourMaxVersion, " +
                    "theirs=${theirs.minVersion}..${theirs.maxVersion})",
            )
            return null
        }

        // Pick the highest mutually-supported version.
        val chosenVersion = overlapMax.toByte()

        // Capability intersection (case-sensitive — capability names are
        // wire constants, not human strings).
        val intersection = LinkedHashSet<String>()
        for (cap in theirs.capabilities) {
            if (cap.isNotEmpty() && ourCapabilities.contains(cap)) {
                intersection.add(cap)
            }
        }

        return PeerCapabilities(
            peerUhid = peerUhid,
            negotiatedVersion = chosenVersion,
            capabilities = intersection,
            implementationVersion = theirs.implementation,
            negotiatedAt = Instant.now(),
        )
    }

    private fun fireIncompatible(peerUhid: String, theirs: HelloPayload, reason: String) {
        logger.warn("Incompatible peer {}: {}", peerUhid, reason)
        onIncompatiblePeer?.invoke(
            IncompatiblePeerEvent(
                peerUhid = peerUhid,
                theirMinVersion = theirs.minVersion,
                theirMaxVersion = theirs.maxVersion,
                ourMinVersion = ourMinVersion,
                ourMaxVersion = ourMaxVersion,
                reason = reason,
            ),
        )
    }

    companion object {
        /** Default capability tags advertised by this implementation. */
        val DEFAULT_CAPABILITIES: Set<String> = linkedSetOf(
            "signal-x3dh",
            "double-ratchet",
            "dtn-custody",
            "sos",
            "voice",
            "stream",
        )

        /** Default implementation banner emitted in our Hello / HelloAck. */
        const val DEFAULT_IMPLEMENTATION: String = "aether-kotlin/1.0.0"
    }
}

File: src/AetherNet.Core/Handshake/HandshakeService.cs
// SPDX-License-Identifier: MIT

using System.Collections.Concurrent;
using System.Text.Json;
using AetherNet.Constants;
using AetherNet.Extensibility;
using AetherNet.Extensibility.Events;
using AetherNet.Protocol;
using AetherNet.Routing;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace AetherNet.Handshake;

/// <summary>
/// Default <see cref="IHandshakeService"/> implementation. Tracks the peers
/// we've Hello'd, the peers we've finished negotiating with, and emits
/// events on completion / incompatibility.
///
/// <para>
/// Wire flow:
/// </para>
/// <code>
/// A → B   Hello       { min:1, max:2, caps:[X,Y,Z], impl:"…" }
/// A ← B   HelloAck    { min:1, max:2, caps:[X,Y],   impl:"…" }
/// </code>
///
/// <para>
/// Negotiation rules:
/// </para>
/// <list type="bullet">
///   <item>Negotiated version = <c>min(ourMax, theirMax)</c>.</item>
///   <item>If <c>min(ourMax,theirMax) &lt; max(ourMin,theirMin)</c> the ranges
///   do not overlap → fire <see cref="IncompatiblePeer"/>, refuse to lock in.</item>
///   <item>Locked-in capability set = <c>ourCaps ∩ theirCaps</c>.</item>
/// </list>
/// </summary>
public sealed class HandshakeService : IHandshakeService
{
    /// <summary>Default capability tags advertised by this implementation.</summary>
    public static readonly IReadOnlySet<string> DefaultCapabilities = new HashSet<string>(StringComparer.Ordinal)
    {
        "signal-x3dh",
        "double-ratchet",
        "dtn-custody",
        "sos",
        "voice",
        "stream",
    };

    /// <summary>Default implementation banner emitted in our Hello/HelloAck.</summary>
    public const string DefaultImplementation = "aether/2";

    private readonly IMeshSender _sender;
    private readonly ILogger<HandshakeService> _logger;
    private readonly byte _ourMinVersion;
    private readonly byte _ourMaxVersion;
    private readonly IReadOnlySet<string> _ourCapabilities;
    private readonly string _ourImplementation;
    private readonly IAetherNetTelemetry? _telemetry;
    private readonly IBiometricProvider _biometricProvider;

    // Peers we've already sent a Hello to, to suppress duplicate sends.
    private readonly ConcurrentDictionary<string, byte> _helloSent = new(StringComparer.Ordinal);

    // Peers we've finished negotiating with.
    private readonly ConcurrentDictionary<string, PeerCapabilities> _negotiated = new(StringComparer.Ordinal);

    public event EventHandler<PeerCapabilities>? PeerNegotiated;
    public event EventHandler<IncompatiblePeerEventArgs>? IncompatiblePeer;

    /// <summary>
    /// Construct a handshake service. Defaults match this codebase: we speak
    /// versions 1..<see cref="ProtocolConstants.CurrentProtocolVersion"/> and
    /// advertise <see cref="DefaultCapabilities"/>.
    /// </summary>
    public HandshakeService(
        IMeshSender sender,
        ILogger<HandshakeService>? logger = null,
        byte? ourMinVersion = null,
        byte? ourMaxVersion = null,
        IReadOnlySet<string>? ourCapabilities = null,
        string? ourImplementation = null,
        IAetherNetTelemetry? telemetry = null,
        IBiometricProvider? biometricProvider = null)
    {
        _sender = sender ?? throw new ArgumentNullException(nameof(sender));
        _logger = logger ?? NullLogger<HandshakeService>.Instance;
        _ourMinVersion = ourMinVersion ?? 1;
        _ourMaxVersion = ourMaxVersion ?? ProtocolConstants.CurrentProtocolVersion;
        if (_ourMinVersion > _ourMaxVersion)
            throw new ArgumentException(
                $"ourMinVersion ({_ourMinVersion}) cannot exceed ourMaxVersion ({_ourMaxVersion}).",
                nameof(ourMinVersion));
        _ourCapabilities = ourCapabilities ?? DefaultCapabilities;
        _ourImplementation = ourImplementation ?? DefaultImplementation;
        _telemetry = telemetry;
        _biometricProvider = biometricProvider ?? NullBiometricProvider.Instance;
    }

    public async Task InitiateAsync(string peerUhid, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(peerUhid);
        if (string.Equals(peerUhid, _sender.LocalUhid, StringComparison.Ordinal)) return;

        // Suppress duplicate Hellos.
        if (!_helloSent.TryAdd(peerUhid, 0))
            return;

        var hello = BuildPacket(PacketType.Hello, peerUhid);
        var delivered = await _sender.SendAsync(hello, peerUhid, cancellationToken).ConfigureAwait(false);
        _logger.LogDebug("Hello sent to {Peer} delivered={Delivered}", peerUhid, delivered);
    }

    public async Task HandleHelloAsync(MeshPacket helloPacket, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(helloPacket);
        if (helloPacket.Type != PacketType.Hello)
            throw new ArgumentException($"Expected Hello, got {helloPacket.Type}", nameof(helloPacket));

        if (string.IsNullOrEmpty(helloPacket.SourceUhid)) return;
        if (string.Equals(helloPacket.SourceUhid, _sender.LocalUhid, StringComparison.Ordinal)) return;

        var theirs = TryDeserialize(helloPacket);
        if (theirs is null)
        {
            _logger.LogWarning("Hello from {Peer} has malformed payload — ignoring",
                helloPacket.SourceUhid);
            return;
        }

        if (!TryNegotiate(helloPacket.SourceUhid, theirs, out var negotiated))
            return; // IncompatiblePeer already fired by TryNegotiate

        _negotiated[helloPacket.SourceUhid] = negotiated;
        PeerNegotiated?.Invoke(this, negotiated);
        _telemetry?.Publish(new AetherNetNodeEvent(
            helloPacket.SourceUhid,
            AetherNetNodeEventKind.Joined,
            new AetherNetNodeHealth(TrustScore: 1.0, IsReachable: true,
                Latency: TimeSpan.Zero, HopCount: 1),
            DateTimeOffset.UtcNow));
        _logger.LogInformation(
            "Hello accepted from {Peer} → version={Ver} caps=[{Caps}] impl={Impl}",
            helloPacket.SourceUhid,
            negotiated.NegotiatedVersion,
            string.Join(",", negotiated.Capabilities),
            negotiated.ImplementationVersion);

        // Reply with HelloAck — even if we already sent them an unprompted Hello,
        // the spec is symmetric and the ack carries our own range/caps.
        var ack = BuildPacket(PacketType.HelloAck, helloPacket.SourceUhid);
        var delivered = await _sender.SendAsync(ack, helloPacket.SourceUhid, cancellationToken).ConfigureAwait(false);
        _logger.LogDebug("HelloAck sent to {Peer} delivered={Delivered}",
            helloPacket.SourceUhid, delivered);
    }

    public Task HandleHelloAckAsync(MeshPacket helloAckPacket, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(helloAckPacket);
        if (helloAckPacket.Type != PacketType.HelloAck)
            throw new ArgumentException($"Expected HelloAck, got {helloAckPacket.Type}", nameof(helloAckPacket));

        if (string.IsNullOrEmpty(helloAckPacket.SourceUhid)) return Task.CompletedTask;
        if (string.Equals(helloAckPacket.SourceUhid, _sender.LocalUhid, StringComparison.Ordinal))
            return Task.CompletedTask;

        var theirs = TryDeserialize(helloAckPacket);
        if (theirs is null)
        {
            _logger.LogWarning("HelloAck from {Peer} has malformed payload — ignoring",
                helloAckPacket.SourceUhid);
            return Task.CompletedTask;
        }

        if (!TryNegotiate(helloAckPacket.SourceUhid, theirs, out var negotiated))
            return Task.CompletedTask; // IncompatiblePeer already fired

        _negotiated[helloAckPacket.SourceUhid] = negotiated;
        PeerNegotiated?.Invoke(this, negotiated);
        _telemetry?.Publish(new AetherNetNodeEvent(
            helloAckPacket.SourceUhid,
            AetherNetNodeEventKind.Joined,
            new AetherNetNodeHealth(TrustScore: 1.0, IsReachable: true,
                Latency: TimeSpan.Zero, HopCount: 1),
            DateTimeOffset.UtcNow));
        _logger.LogInformation(
            "HelloAck received from {Peer} → version={Ver} caps=[{Caps}] impl={Impl}",
            helloAckPacket.SourceUhid,
            negotiated.NegotiatedVersion,
            string.Join(",", negotiated.Capabilities),
            negotiated.ImplementationVersion);

        return Task.CompletedTask;
    }

    public Task<PeerCapabilities?> GetPeerCapabilitiesAsync(
        string peerUhid,
        CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(peerUhid);
        return Task.FromResult(_negotiated.TryGetValue(peerUhid, out var caps) ? caps : null);
    }

    public Task RenegotiateAsync(string peerUhid, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(peerUhid);
        _negotiated.TryRemove(peerUhid, out _);
        _helloSent.TryRemove(peerUhid, out _);
        _logger.LogInformation("Cleared cached capabilities for {Peer}; next contact will re-Hello", peerUhid);
        return Task.CompletedTask;
    }

    public IReadOnlyList<PeerCapabilities> GetAllNegotiated()
        => _negotiated.Values.ToArray();

    /// <inheritdoc/>
    public async Task<BiometricVerificationResult> VerifyCoPresenceAsync(
        byte[]            localFaceFrameRgbHwc,
        int               width,
        int               height,
        FaceEmbedding     referenceEmbedding,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(referenceEmbedding);

        if (!_biometricProvider.IsAvailable)
        {
            _logger.LogDebug("VerifyCoPresence: biometric provider unavailable — returning Failed");
            return BiometricVerificationResult.Failed;
        }

        // Detect the dominant face in the live frame.
        var faces = await _biometricProvider.DetectAsync(
            localFaceFrameRgbHwc, width, height, maxFaces: 1, cancellationToken)
            .ConfigureAwait(false);

        if (faces.Count == 0)
        {
            _logger.LogDebug("VerifyCoPresence: no face detected in live frame — returning Failed");
            return BiometricVerificationResult.Failed;
        }

        var detected = faces[0];
        if (!detected.IsConfident)
        {
            _logger.LogDebug(
                "VerifyCoPresence: detection confidence {Score:F2} below threshold 0.50 — returning Failed",
                detected.DetectionScore);
            return BiometricVerificationResult.Failed;
        }

        // Compare the detected face to the peer's reference embedding.
        var verifyResult = await _biometricProvider.VerifyAsync(
            referenceEmbedding, detected.Embedding, cancellationToken: cancellationToken)
            .ConfigureAwait(false);

        _logger.LogDebug(
            "VerifyCoPresence: verified={Verified} similarity={Similarity:F3}",
            verifyResult.Verified, verifyResult.Similarity);

        return verifyResult;
    }

    /// <summary>
    /// Backward-compat: install a "v1, no caps" record for a peer that never
    /// replied to our Hello within the timeout window. Hosts call this from
    /// their own timer / heartbeat loop. Idempotent — if the peer has since
    /// replied with a HelloAck, the existing record wins.
    /// </summary>
    public void AssumeLegacyV1(string peerUhid)
    {
        ArgumentException.ThrowIfNullOrEmpty(peerUhid);
        if (string.Equals(peerUhid, _sender.LocalUhid, StringComparison.Ordinal)) return;

        var fallback = new PeerCapabilities(
            peerUhid,
            NegotiatedVersion: 1,
            Capabilities: new HashSet<string>(StringComparer.Ordinal),
            ImplementationVersion: string.Empty,
            NegotiatedAt: DateTimeOffset.UtcNow);

        var added = _negotiated.GetOrAdd(peerUhid, fallback);
        if (ReferenceEquals(added, fallback))
        {
            PeerNegotiated?.Invoke(this, fallback);
            _logger.LogWarning(
                "No HelloAck from {Peer} after timeout — assuming protocol v1 / no advertised capabilities",
                peerUhid);
        }
    }

    private MeshPacket BuildPacket(PacketType type, string destinationUhid)
    {
        var payload = new HelloPayload
        {
            MinVersion = _ourMinVersion,
            MaxVersion = _ourMaxVersion,
            Capabilities = _ourCapabilities.ToList(),
            Implementation = _ourImplementation,
        };

        return new MeshPacket
        {
            Type = type,
            SourceUhid = _sender.LocalUhid,
            DestinationUhid = destinationUhid,
            Ttl = 1, // direct hop only — handshake never relays
            Priority = 0,
            ProtocolVersion = _ourMaxVersion,
            Payload = JsonSerializer.SerializeToUtf8Bytes(payload, HelloPayloadJson.Options),
        };
    }

    private HelloPayload? TryDeserialize(MeshPacket packet)
    {
        if (packet.Payload is null || packet.Payload.Length == 0) return null;
        try
        {
            return JsonSerializer.Deserialize<HelloPayload>(packet.Payload, HelloPayloadJson.Options);
        }
        catch (JsonException ex)
        {
            _logger.LogWarning(ex, "Handshake payload from {Peer} could not be parsed", packet.SourceUhid);
            return null;
        }
    }

    private bool TryNegotiate(string peerUhid, HelloPayload theirs, out PeerCapabilities negotiated)
    {
        negotiated = null!;

        if (theirs.MinVersion > theirs.MaxVersion)
        {
            _logger.LogWarning(
                "Handshake from {Peer} announces inverted range min={Min} > max={Max} — refusing",
                peerUhid, theirs.MinVersion, theirs.MaxVersion);
            FireIncompatible(peerUhid, theirs, "inverted version range");
            return false;
        }

        // Overlap check: highest min must be ≤ lowest max.
        var overlapMin = Math.Max(_ourMinVersion, theirs.MinVersion);
        var overlapMax = Math.Min(_ourMaxVersion, theirs.MaxVersion);
        if (overlapMin > overlapMax)
        {
            FireIncompatible(peerUhid, theirs,
                $"no version overlap (ours={_ourMinVersion}..{_ourMaxVersion}, theirs={theirs.MinVersion}..{theirs.MaxVersion})");
            return false;
        }

        // Pick the highest mutually-supported version.
        var chosenVersion = (byte)overlapMax;

        // Capability intersection (case-sensitive, ordinal — capability names
        // are wire constants, not human strings).
        var theirCaps = theirs.Capabilities ?? new List<string>();
        var intersection = new HashSet<string>(StringComparer.Ordinal);
        foreach (var cap in theirCaps)
        {
            if (!string.IsNullOrEmpty(cap) && _ourCapabilities.Contains(cap))
                intersection.Add(cap);
        }

        negotiated = new PeerCapabilities(
            peerUhid,
            chosenVersion,
            intersection,
            theirs.Implementation ?? string.Empty,
            DateTimeOffset.UtcNow);
        return true;
    }

    private void FireIncompatible(string peerUhid, HelloPayload theirs, string reason)
    {
        _logger.LogWarning(
            "Incompatible peer {Peer}: {Reason}",
            peerUhid, reason);
        IncompatiblePeer?.Invoke(this,
            new IncompatiblePeerEventArgs(
                peerUhid,
                theirs.MinVersion,
                theirs.MaxVersion,
                _ourMinVersion,
                _ourMaxVersion,
                reason));
    }
}
